File: Packages/NMP/Sources/NMP/Following.swift
import Combine
import NMPFFI

public enum NMPFollowRelationship: Sendable, Hashable {
    case unknown
    case notFollowing
    case following

    init(_ ffi: FfiFollowRelationship) {
        switch ffi {
        case .unknown: self = .unknown
        case .notFollowing: self = .notFollowing
        case .following: self = .following
        }
    }
}

/// Source evidence for the live relationship projection. It does not gate
/// follow/unfollow: NMP can write cached state or create the first list while
/// relay truth is incomplete. `.ready` is not global Nostr completeness.
public enum NMPFollowAvailability: Sendable, Hashable {
    case signedOut
    case acquiring
    case ready
    case noContactList
    case cachedOnly
    case sourceUnavailable

    init(_ ffi: FfiFollowAvailability) {
        switch ffi {
        case .signedOut: self = .signedOut
        case .acquiring: self = .acquiring
        case .ready: self = .ready
        case .noContactList: self = .noContactList
        case .cachedOnly: self = .cachedOnly
        case .sourceUnavailable: self = .sourceUnavailable
        }
    }
}

public struct NMPFollowingSnapshot: Sendable, Hashable {
    public let currentPubkey: String?
    public let target: String
    public let relationship: NMPFollowRelationship
    public let availability: NMPFollowAvailability
    public let baseEventID: String?

    public init(
        currentPubkey: String?,
        target: String,
        relationship: NMPFollowRelationship,
        availability: NMPFollowAvailability,
        baseEventID: String?
    ) {
        self.currentPubkey = currentPubkey
        self.target = target
        self.relationship = relationship
        self.availability = availability
        self.baseEventID = baseEventID
    }

    init(_ ffi: FfiFollowSnapshot) {
        self.init(
            currentPubkey: ffi.currentPubkey,
            target: ffi.target,
            relationship: NMPFollowRelationship(ffi.relationship),
            availability: NMPFollowAvailability(ffi.availability),
            baseEventID: ffi.baseEventId
        )
    }

    public static func initial(target: String) -> Self {
        Self(
            currentPubkey: nil,
            target: target,
            relationship: .unknown,
            availability: .acquiring,
            baseEventID: nil
        )
    }
}

/// A typed follow/unfollow action was refused before ordinary receipt
/// custody. `invalidTarget` is the one refusal this boundary adds: `target`
/// crosses FFI as a caller-typed hex string.
public enum FollowActionError: Error, Sendable, Equatable {
    case invalidTarget(got: String)
    case automaticRoutingUnavailable
    case signedOut
    case engineClosed
    case publishRefused(reason: String)

    init(_ ffi: FfiFollowActionError) {
        switch ffi {
        case .InvalidTarget(let got): self = .invalidTarget(got: got)
        case .AutomaticRoutingUnavailable: self = .automaticRoutingUnavailable
        case .SignedOut: self = .signedOut
        case .EngineClosed: self = .engineClosed
        case .PublishRefused(let reason): self = .publishRefused(reason: reason)
        }
    }
}

/// Live relationship state over NMP's ordinary reactive kind:3 demand
/// (#680). A pull-based `AsyncSequence` over `NmpFollowStream` -- each
/// snapshot is the complete self-contained relationship state (latest-wins),
/// so no coalescer is needed. Termination-tied teardown like `NMPQuery`.
public struct NMPFollowingObservation: AsyncSequence, Sendable {
    public typealias Element = NMPFollowingSnapshot

    private let handle: NmpFollowStream
    private let iteratorGate = NMPPullIteratorGate()

    init(engine: NmpEngineProtocol, target: String) throws {
        self.handle = try nmpRethrowing {
            try engine.observeFollowing(target: target)
        }
    }

    public func makeAsyncIterator() -> Iterator {
        let core = NMPPullIteratorCore(handle: handle, iteratorGate: iteratorGate) { snapshot in
            NMPFollowingSnapshot(snapshot)
        }
        return Iterator(core: core)
    }

    public struct Iterator: AsyncIteratorProtocol {
        let core: NMPPullIteratorCore<NmpFollowStream, NMPFollowingSnapshot>

        public mutating func next() async throws -> NMPFollowingSnapshot? {
            try await core.next()
        }
    }

    public func cancel() {
        handle.cancel()
    }
}

extension NMPEngine {
    /// Observe whether the current account follows `target`. This is NMP's
    /// protocol projection, not an app-maintained boolean.
    public func observeFollowing(_ target: String) throws -> NMPFollowingObservation {
        try NMPFollowingObservation(engine: ffi, target: target)
    }

    /// Ask NMP to follow `target` through the ordinary durable write and
    /// receipt lifecycle. NMP applies it immediately to the best cached
    /// contact list, or to NIP-02's complete empty list when no source
    /// exists, and reapplies it if a newer relay source arrives. Either a
    /// truthful immediate `FollowActionError`, or the same `Receipt` every
    /// other write returns.
    public func follow(_ target: String) throws -> Receipt {
        try followReceipt { try ffi.follow(target: target) }
    }

    /// The inverse of `follow(_:)`, with the same durable operation and
    /// ordinary receipt guarantees.
    public func unfollow(_ target: String) throws -> Receipt {
        try followReceipt { try ffi.unfollow(target: target) }
    }

    private func followReceipt(_ action: () throws -> NmpReceiptStream) throws -> Receipt {
        do {
            return Receipt(handle: try action())
        } catch let error as FfiFollowActionError {
            throw FollowActionError(error)
        }
    }
}

/// Bindable convenience over the two NMP APIs above. It owns no NIP-02
/// logic: snapshots and action statuses are copied directly from Rust; the
/// only local state is observation/task lifecycle for SwiftUI.
@MainActor
public final class NMPFollowing: ObservableObject {
    public let target: String

    @Published public private(set) var snapshot: NMPFollowingSnapshot
    @Published public private(set) var actionStatus: WriteFact?
    @Published public private(set) var isActing = false

    private let engine: NMPEngine
    private var desiredFollowing: Bool?
    private nonisolated(unsafe) var observationTask: Task<Void, Never>?
    private nonisolated(unsafe) var actionTask: Task<Void, Never>?

    public init(engine: NMPEngine, target: String) throws {
        self.engine = engine
        self.target = target
        self.snapshot = .initial(target: target)
        let observation = try engine.observeFollowing(target)
        observationTask = Task { [weak self] in
            do {
                for try await snapshot in observation {
                    guard !Task.isCancelled else { return }
                    self?.snapshot = snapshot
                    self?.finishWhenCanonicalStateMatches(snapshot)
                }
            } catch {
                // The observation ended (withdrawal / single-consumer misuse);
                // stop updating. NMP surfaces no capacity error here (#680).
            }
        }
    }

    public var canToggle: Bool {
        snapshot.currentPubkey != nil
            && snapshot.relationship != .unknown
            && !isActing
    }

    public func follow() {
        start(desiredFollowing: true)
    }

    public func unfollow() {
        start(desiredFollowing: false)
    }

    public func toggle() {
        guard canToggle else { return }
        switch snapshot.relationship {
        case .following: unfollow()
        case .notFollowing: follow()
        case .unknown: break
        }
    }

    /// The single action a connected control forwards. NMP retains and
    /// reapplies the operation; the UI owns no stale-base retry policy.
    public func performPrimaryAction() {
        guard canToggle else { return }
        toggle()
    }

    private func start(desiredFollowing: Bool) {
        guard !isActing else { return }
        let receipt: Receipt
        do {
            receipt = try (desiredFollowing ? engine.follow(target) : engine.unfollow(target))
        } catch {
            // A truthful immediate refusal (FollowActionError): nothing
            // entered custody, so there is no stream to follow.
            self.desiredFollowing = nil
            self.isActing = false
            return
        }
        self.desiredFollowing = desiredFollowing
        self.isActing = true
        self.actionStatus = nil
        actionTask?.cancel()
        actionTask = Task { [weak self] in
            do {
                for try await fact in receipt.status {
                    guard !Task.isCancelled else { return }
                    self?.accept(fact)
                }
            } catch {
                // The receipt stream ended abnormally; leave the last delivered
                // status in place (no capacity error exists to surface, #680).
            }
        }
    }

    private func accept(_ fact: WriteFact) {
        actionStatus = fact
        if case .outcome(.refused) = fact {
            isActing = false
            desiredFollowing = nil
        } else if case .signing(.refused) = fact {
            isActing = false
            desiredFollowing = nil
        }
    }

    private func finishWhenCanonicalStateMatches(_ snapshot: NMPFollowingSnapshot) {
        guard let desiredFollowing else { return }
        let matches = desiredFollowing
            ? snapshot.relationship == .following
            : snapshot.relationship == .notFollowing
        if matches {
            isActing = false
            self.desiredFollowing = nil
        }
    }

    deinit {
        observationTask?.cancel()
        actionTask?.cancel()
    }
}

File: Packages/NMPKotlin/src/main/kotlin/com/nmp/sdk/Following.kt
// NMP's optional NIP-02 following resource/action, in ergonomic Kotlin
// shape. Mirrors Following.swift's `NMPEngine.observeFollowing`/`follow`/
// `unfollow` PUBLIC semantics exactly: same states, same typed failures,
// same "the action returns immediately with a status stream" contract. No
// contact-list parsing, replacement composition, or readiness policy lives
// on this side of the FFI boundary -- this file only mirrors Rust-owned
// state and drains Rust-owned streams, exactly like `nmp-ffi/src/nip02.rs`'s
// own header comment says of the Rust side.
//
// SCOPE NOTE: Following.swift also defines `NMPFollowing`, a `@MainActor`
// `ObservableObject` that bundles `canToggle`/`offersAnotherAttempt`/
// `toggle()`/`performPrimaryAction()` local UI-state bookkeeping on top of
// the two APIs below. That class is SwiftUI-specific presentation sugar,
// exactly the same shape as `Observable.swift`'s `NMPQuerySnapshot` and
// `NMPDiagnosticsSnapshotObserver` -- and this codebase's established
// precedent (see that file's own header) is to NOT port those `@Observable`
// convenience wrappers to Kotlin: `Query.kt`/`DiagnosticsQuery.kt` stop at
// the `Flow`-returning primary API and leave the ObservableObject sugar
// unbuilt on this platform. This file follows that same precedent and stops
// at the primary API; it does not invent a StateFlow/ViewModel-shaped
// counterpart to `NMPFollowing`. If Android callers need the toggle/retry
// state machine ported too, that is a separate, explicit follow-up -- not
// guessed here.
package com.nmp.sdk

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import uniffi.nmp_ffi.FfiFollowActionFailure
import uniffi.nmp_ffi.FfiFollowActionStatus
import uniffi.nmp_ffi.FfiFollowAvailability
import uniffi.nmp_ffi.FfiFollowRelationship
import uniffi.nmp_ffi.FfiFollowSnapshot
import uniffi.nmp_ffi.NmpEngineInterface
import uniffi.nmp_ffi.NmpFollowActionStream

/** The current account's relationship to a `target` pubkey, as NMP's own
 * kind:3 projection sees it right now (`FfiFollowRelationship` mirror). */
enum class FollowRelationship {
    Unknown,
    NotFollowing,
    Following,
    ;

    companion object {
        fun from(ffi: FfiFollowRelationship): FollowRelationship =
            when (ffi) {
                FfiFollowRelationship.UNKNOWN -> Unknown
                FfiFollowRelationship.NOT_FOLLOWING -> NotFollowing
                FfiFollowRelationship.FOLLOWING -> Following
            }
    }
}

/** Source evidence for the live relationship projection
 * (`FfiFollowAvailability` mirror). It does not gate follow/unfollow: NMP can
 * write cached state or create the first list while relay truth is incomplete.
 * `Ready` is not global Nostr completeness. */
enum class FollowAvailability {
    SignedOut,
    Acquiring,
    Ready,
    NoContactList,
    CachedOnly,
    SourceUnavailable,
    ;

    companion object {
        fun from(ffi: FfiFollowAvailability): FollowAvailability =
            when (ffi) {
                FfiFollowAvailability.SIGNED_OUT -> SignedOut
                FfiFollowAvailability.ACQUIRING -> Acquiring
                FfiFollowAvailability.READY -> Ready
                FfiFollowAvailability.NO_CONTACT_LIST -> NoContactList
                FfiFollowAvailability.CACHED_ONLY -> CachedOnly
                FfiFollowAvailability.SOURCE_UNAVAILABLE -> SourceUnavailable
            }
    }
}

/** One pushed state of the current account's relationship to `target`
 * (`FfiFollowSnapshot` mirror). Delivered by `NMPEngine.observeFollowing`,
 * pushed reactively, never polled. */
data class FollowingSnapshot(
    val currentPubkey: String?,
    val target: String,
    val relationship: FollowRelationship,
    val availability: FollowAvailability,
    val baseEventId: String?,
) {
    companion object {
        fun from(ffi: FfiFollowSnapshot): FollowingSnapshot =
            FollowingSnapshot(
                currentPubkey = ffi.currentPubkey,
                target = ffi.target,
                relationship = FollowRelationship.from(ffi.relationship),
                availability = FollowAvailability.from(ffi.availability),
                baseEventId = ffi.baseEventId,
            )

        /** The pre-acquisition placeholder a caller may render before the
         * first real snapshot arrives (mirrors
         * `NMPFollowingSnapshot.initial(target:)`). Kotlin's cold `Flow`
         * -- unlike Swift's synchronously-constructed `struct` -- has no
         * value at all until the flow is collected and the engine has
         * delivered one, so a caller that wants an immediate placeholder value
         * (e.g. to seed a `MutableStateFlow` before `collect` starts) uses
         * this explicitly; `observeFollowing` itself only ever emits real,
         * engine-sourced snapshots. */
        fun initial(target: String): FollowingSnapshot =
            FollowingSnapshot(
                currentPubkey = null,
                target = target,
                relationship = FollowRelationship.Unknown,
                availability = FollowAvailability.Acquiring,
                baseEventId = null,
            )
    }
}

/** Every typed way NMP's follow/unfollow action can end without changing
 * the relationship (`FfiFollowActionFailure` mirror). */
sealed class FollowActionFailure {
    data class InvalidTarget(val got: String) : FollowActionFailure()

    object SignedOut : FollowActionFailure()

    object EngineClosed : FollowActionFailure()

    object ReceiptUnavailable : FollowActionFailure()

    companion object {
        fun from(ffi: FfiFollowActionFailure): FollowActionFailure =
            when (ffi) {
                is FfiFollowActionFailure.InvalidTarget -> InvalidTarget(ffi.got)
                is FfiFollowActionFailure.SignedOut -> SignedOut
                is FfiFollowActionFailure.EngineClosed -> EngineClosed
                is FfiFollowActionFailure.ReceiptUnavailable -> ReceiptUnavailable
            }
    }
}

/** A thin projection of the ordinary receipt created by a typed
 * `follow`/`unfollow` action. Successful actions contain only canonical
 * [WriteFact] values; immediate typed refusal is the sole non-receipt case. */
sealed class FollowActionStatus {
    data class Receipt(val id: ULong, val status: WriteFact) : FollowActionStatus()

    data class Failed(val failure: FollowActionFailure) : FollowActionStatus()

    companion object {
        fun from(ffi: FfiFollowActionStatus): FollowActionStatus =
            when (ffi) {
                is FfiFollowActionStatus.Receipt ->
                    Receipt(ffi.receiptId, WriteFact.from(ffi.status))
                is FfiFollowActionStatus.Failed -> Failed(FollowActionFailure.from(ffi.failure))
            }
    }
}

/** A typed follow/unfollow action's ordinary receipt projection. The stable
 * receipt id arrives inside [FollowActionStatus.Receipt]. */
data class FollowAction(val status: Flow<FollowActionStatus>)

/** Observe whether the current account follows [target] through the
 * NMP-owned NIP-02 resource (mirrors `NMPEngine.observeFollowing`). This is
 * NMP's protocol projection, not an app-maintained boolean.
 *
 * Each element is the full current [FollowingSnapshot] -- latest-wins,
 * never a growing backlog: the engine's latest-state mailbox conflates
 * intermediate snapshots for a slow collector (#680), so the wrapper is a
 * thin pull loop over `NmpFollowStream.next()`. Demand teardown is
 * collection-scope-tied via `handle.cancel()` in a `finally`, identical
 * reasoning to `Query.kt`'s header. */
fun observeFollowing(engine: NmpEngineInterface, target: String): Flow<FollowingSnapshot> =
    flow {
        val handle = nmpRethrowing { engine.observeFollowing(target) }
        try {
            while (true) {
                val snapshot = nmpRethrowingAsync { handle.next() } ?: break
                emit(FollowingSnapshot.from(snapshot))
            }
        } finally {
            handle.cancel()
        }
    }

/** Shared pull loop for the ordinary receipt facts projected by a typed
 * `follow`/`unfollow` action. Teardown is collection-scope-tied via
 * `handle.cancel()`. */
private fun followActionFlow(open: () -> NmpFollowActionStream): Flow<FollowActionStatus> =
    flow {
        val handle = nmpRethrowing { open() }
        try {
            while (true) {
                val status = nmpRethrowingAsync { handle.next() } ?: break
                emit(FollowActionStatus.from(status))
            }
        } finally {
            handle.cancel()
        }
    }

/** Ask NMP to follow [target] (mirrors `NMPEngine.follow`). NMP immediately
 * applies one durable semantic operation to the best cached contact list, or
 * to NIP-02's complete empty list when no source exists. If a newer relay
 * source arrives, NMP reapplies the same operation while retaining the same
 * receipt. The caller only observes [FollowAction.status]. A missing automatic
 * route provider is refused before custody. An invalid [target] surfaces as
 * `FollowActionStatus.Failed(FollowActionFailure.InvalidTarget)` on the
 * stream, not as a synchronous exception). */
fun follow(engine: NmpEngineInterface, target: String): FollowAction =
    FollowAction(followActionFlow { engine.follow(target) })

/** The inverse of [follow], with the same durable semantic operation and
 * ordinary receipt guarantees (mirrors `NMPEngine.unfollow`). */
fun unfollow(engine: NmpEngineInterface, target: String): FollowAction =
    FollowAction(followActionFlow { engine.unfollow(target) })
